### PR TITLE
inverse model & atari hyperparam changes 

### PR DESCRIPTION
**Pull request by zhouzypaul**

### CODE REVIEW

**Comment by zhouzypaul** [comment]
@camall3n by the way, torch is still deterministic on this branch.

**Comment by zhouzypaul** [comment]
- change hyperparams according to other repos
- change epsilon decay schedule from linear to exponential 
- don't save loss info by default 
- disable clip_rewards and episodic_life in test_env
- add inverse model and train it concurrently with DQN

File: actgen/agents/dqn.py [hunk 1]
```diff
@@ -99,6 +108,10 @@ def _get_action_similarities(self, batch):
             for i, acted in enumerate(action_taken):
                 all_duplicate_actions = self.get_duplicate_actions_fn(acted)
                 similarity_mat[i, all_duplicate_actions] = 1  # update each row in similarity_mat
+        elif self.params['inv_model']:
+            action_probs = self.inverse_predictor.predict(batch, encoder=self.q.encoder)  # (batch_size, n_actions)
+            clipped_action_probs = torch.clip(action_probs, min=self.params['inv_clip_threshold'])  # (batch_size, n_actions)
+            similarity_mat = clipped_action_probs
```
Comment (camall3n) [inline actgen/agents/dqn.py]: What are you using this `clip` operation for?

File: actgen/agents/dqn.py [hunk 2]
```diff
@@ -74,16 +80,19 @@ def update(self):
...
 
     def _get_epsilon(self, testing=False):
         if testing:
             epsilon = self.params['epsilon_during_eval']
         else:
-            alpha = (len(self.replay) - self.params['replay_warmup_steps']) / self.params['epsilon_decay_period']
-            alpha = np.clip(alpha, 0, 1)
-            epsilon = self.params['epsilon_final'] * alpha + 1 * (1 - alpha)
+            epsilon = (self.params['epsilon_final'] + (1 - self.params['epsilon_final'])
+                 * math.exp(-1. * self.n_training_steps / self.params['epsilon_decay_period']))
```
Comment (camall3n) [inline actgen/agents/dqn.py:96]: This was the switch to an exponential epsilon schedule that we talked about, right?

File: actgen/agents/dqn.py (hunk 1 above)
Comment (zhouzypaul) [reply actgen/agents/dqn.py]: I don't quite remember if we agreed on how we would use the output of the inverse model, so what I'm doing now is:

- the output of the inverse model is a categorical distribution that sums to 1 (e.g. 0.05, 0.55, 0.40 for 3-action-env)
- if the probability for a certain action is smaller than `inv_clip_threshold=0.1`, then the action is not really likely to be the true action that led to the transition, so we don't use that similarity_score as a approximate oracle. 

The `clip` operation is a bug, I just realized. What I'm looking for was `torch.where(x <= 0.1, 0, x)`

File: actgen/agents/dqn.py (hunk 2 above)
Comment (zhouzypaul) [reply actgen/agents/dqn.py:96]: yes!

File: actgen/agents/dqn.py (hunk 1 above)
Comment (camall3n) [reply actgen/agents/dqn.py]: Okay, let's discuss how we want to use the inv model at the meeting today

File: actgen/models/inverse.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            self.mean_linear = torch.nn.Linear(params['inv_layer_size'], n_actions)
+            self.log_std_linear = torch.nn.Linear(params['inv_layer_size'], n_actions)
+            self.log_sig_min = params['inv_log_std_min']
+            self.log_sig_max = params['inv_log_std_max']
+
+    def forward(self, z0, z1):
+        """
+        Compute the action distribution for two subsequent states.
+
+        If the action space is discrete, the output will be log probabilities,
```
Comment (zhouzypaul) [inline actgen/models/inverse.py:38]: @camall3n the comments here say the output will be log probabilities, but it looks like `self.log_pr_linear` is just a linear layer, where is the `log` coming from?

Comment (camall3n) [reply actgen/models/inverse.py:38]: This is what people typically refer to as "logits", i.e. the input to a softmax layer. Since softmax applies an exponential to produce probabilities, the inputs are effectively log-probabilities.

Comment (camall3n) [reply actgen/models/inverse.py:38]: Note that there is no softmax layer here, so if you want normalized probabilities, you need to send the `log_pr_linear` through a softmax

File: actgen/wrappers/remove_redundant_actions.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+	for action in range(5):
+		assert action in env.action_space
+	assert 9 not in env.action_space
+	assert 5 not in env.action_space
+	assert 6 not in env.action_space
+
+	assert env.action_space.n == 5
+
+	assert 0 <= env.action_space.sample() <= 4
```
Comment (camall3n) [inline actgen/wrappers/remove_redundant_actions.py]: These kinds of randomized test cases worry me a little bit. I'd rather have a for loop that checks each value, rather than relying on randomness.

When you have a randomized test case, it can appear to pass when in fact it sometimes fails, and if we don't notice it fails until a later time, we might incorrectly think some new code is the reason for the failure, instead of knowing it was broken all along.

Comment (camall3n) [reply actgen/wrappers/remove_redundant_actions.py]: Here, we can probably just check that `0 in env.action_space`, `4 in env.action_space`, and maybe that `env` and `gym.make('MsPacmanNoFrameskip-v4')` have the same action_space type (i.e. `Discrete`).

**Comment by camall3n** [comment]
I would just do something like `softmax(inverse_model(z, a, zp))` if you need probabilities.

File: actgen/agents/dqn.py [hunk 5]
```diff
@@ -99,6 +109,18 @@ def _get_action_similarities(self, batch):
...
+            # don't use the inverse model in early stages of training
+            if self.n_training_steps < self.params['inv_warmup_steps']:
+                return similarity_mat
+            # use the inverse model
+            action_probs = self.inverse_predictor.predict(batch, encoder=self.q.encoder)  # (batch_size, n_actions)
+            # scale the probabilities linearly so that the actual action taken has probability 1
+            scale = 1 / extract(action_probs, action_taken, idx_dim=-1).view((self.params['batch_size'], 1))
+            scaled_action_probs = scale * action_probs
+            # those with probabilities lower than 0.5 are deemed none-similar actions
+            clipped_action_probs = torch.where(scaled_action_probs < self.params['inv_clip_threshold'], 0, scaled_action_probs)
```
Comment (camall3n) [inline actgen/agents/dqn.py]: If this doesn't work, let's try without rescaling and clipping. Instead we just modify the acted action similarity to equal 1.0.

**Comment by zhouzypaul** [comment]
new changes: 
small bug fixes and adding support for 6 atari games.

File: actgen/plot.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import os
 import argparse
 import logging
+from pathlib import Path
```
Comment (camall3n) [inline actgen/plot.py:4]: It looks like we aren't using this import

**Review by camall3n — APPROVED** [review]
